jdbi-postgres at 3.7.0 is incompatible with pooled connection from tomcat

Hello,

pull request #1425 aka "add support for postgres custom types", has an unfortunate side effect of breaking jdbi-postgres 3.7.0 when JDBI is used in a Tomcat servlet, with Tomcat doing connection pooling for the application. The issue is with the new method, PostgresPlugin::customizeHandle, which attempts to unwrap the connection to a PGConnection instance, which is now a concrete instance that gets defined locally via direct dependency to pg JDBC library 42.2.4 from jdbi-postgres.

The issue with connection pooling is that Tomcat will have loaded its own Pg JDBC driver in order to do the connection pooling, and it will be serving instance of that via its DataSource interface. As Tomcat gives a custom classloader to every application it is serving, applications that also load a copy of Pg JDBC driver will get a new, unrelated copy of the database driver. Despite the class names are the same, the classes themselves are not, and thus any casting of Tomcat's PGConnection to application's PGConnection is impossible. This is true even if the Pg jar files used by Tomcat and application are exactly the same.

To solve the issue, I think it would be best to only depend on the shared Connection interface, and maybe use reflection to access the required methods of a true PGConnection. The issue is identified by this exception:

Caused by: java.sql.SQLException: Cannot unwrap to org.postgresql.PGConnection
	at org.postgresql.jdbc.PgConnection.unwrap(PgConnection.java:1494)
	at org.apache.tomcat.dbcp.dbcp2.DelegatingConnection.unwrap(DelegatingConnection.java:842)
	at org.apache.tomcat.dbcp.dbcp2.DelegatingConnection.unwrap(DelegatingConnection.java:842)
	at org.jdbi.v3.postgres.PostgresPlugin.lambda$customizeHandle$0(PostgresPlugin.java:150)
	at org.jdbi.v3.core.internal.exceptions.Unchecked.lambda$supplier$3(Unchecked.java:66)
	at org.jdbi.v3.postgres.PostgresPlugin.customizeHandle(PostgresPlugin.java:150)
	at org.jdbi.v3.core.Jdbi.open(Jdbi.java:311)
	... 27 more

It is possible to workaround this issue by using JDBI 3.7.0 otherwise, but selecting version 3.6.0 of jdbi-postgres.

--- a/RELEASE_NOTES.md
+++ b/RELEASE_NOTES.md
@@ -1,6 +1,8 @@
 #3.7.1
 - New Features
   - Initial Java Module System support via Automatic-Module-Name
+- Improvements
+  - Postgres JDBC driver is now <scope>provided</scope> to avoid fighting with servlet containers.
 
 #3.7.0
 - New Features

--- a/postgres/pom.xml
+++ b/postgres/pom.xml
@@ -57,6 +57,7 @@
         <dependency>
             <groupId>org.postgresql</groupId>
             <artifactId>postgresql</artifactId>
+            <scope>provided</scope>
         </dependency>
 
         <dependency>

